arch/x86/crc32_fold_pclmulqdq.c

```diff
@@ -462,4 +462,11 @@ Z_INTERNAL uint32_t crc32_fold_final_pclmulqdq(crc32_fold *crc) {
     return crc->value;
 }
 
+uint32_t crc32_pclmulqdq(uint32_t crc32, const unsigned char* buf, uint64_t len) {
```

Comment:
I don't think this will work right because it doesn't do anything with `crc32` value.

Reply:
Pretty sure it does?  The return statement finalizes the CRC value by folding up the 4 xmm registers to a single sum, no?

Reply:
I wonder if `crc32_combine` can be used to combine the initial value with the final value. OR if there is a way to load the initial value into our `crc32_fold..` stuff.

Reply:
The `crc32` variable is unused in the function..

Reply:
In the benchmarks I had to add some kind of `supports_initial_value` in order to benchmark it... so it doesn't support having an initial `crc32` value..

Reply:
ahh, you are correct there, if it's non-zero that will be a problem.  Yes we should be able to feed it after the "reset" function.

Reply:
Ok, that last push should fix that.

Reply:
I'm doing some tests but I don't think it is possible to just xor the two values..

Reply:
Combining two CRC32 values need at least length of one of the original byte streams... It might be better to use the existing combine function instead...

Reply:
crc32_combine functions look like they also need the length..

Reply:
`crc32_combine` needs the length as it adds number of zeroes between the two byte streams, so they are of equal length. But here we already know how long the second byte stream is...

Reply:
Ahhh, incidentally I'm getting segfaults on the unit tests.  Weirdly, on an unaligned move.  For some reason it does not seem well behaved when unaligned, which is very strange.  

I'll call the combine function instead (though, it did seem like in my testing the simple XOR'ing was correct).

Reply:
@mtl1979 weird that the comment for `crc32_combine` mentions a variable `len1` but only exists `len2`.

Reply:
@KungFuJesus I had mentioned previously benchmarks, I was wrong. It was in the google unit tests PR that I checked for `works_with_initial_value`. Also I have pushed a commit which should fix some cpu feature checking.

Reply:
@nmoinvaz I haven't read the documentation in a while... I just know how combing two crc32 values is supposed to work from reading the source code...

Reply:
@mtl1979 you are right, it works like this.. probably needs to be cleaned up..
<img width="368" alt="image" src="https://user-images.githubusercontent.com/1364432/153105682-5db2088b-5c1e-4143-8125-d8d4a754346d.png">

Reply:
@nmoinvaz AFAIK it doesn't need length of first byte stream as crc32 of initial zeroes is same as resetting the state, so basically has no effect.

Reply:
@KungFuJesus in my crc32 shim I had to have an extra buffer to write to because I think it is passing in constant strings and such as `buf`.

Reply:
Ah well it only matters for writing. That makes a good deal of sense, since constant strings are mapped read only in the binary.

Perhaps it makes sense then to write separate non-copying versions? I suppose we _could_ templatize things such that a macro compiled the copy in or not, I'm not sure if that's much better as far as maintainability.

Off topic but, I had half a mind to write out a similar sort of thing for the Adler checksums where it's used in deflate (roll the values through the checksum while they're in cache or registers) but doing that effectively without making ridiculously long function bodies would require us templating it out like we do with some of the functions. Since we're blistering fast when fitting into cache, avoiding the trip to memory entirely would probably have substantial gains at the cost of writing out inlined versions of sections of the function so that we can wrap deflate into the checksum effectively (maximizing ILP and all of that).

Reply:
So, the crc32_combine function works but when compiling in zlib compatibility mode it complains about the signature being different.  I obviously can cast it down to a 32 bit integer but, is there a cleaner way?

Reply:
Ooof, that crc32_combine function is also not an insignificant amount of the computation time.  It doesn't correlate to a ton of extra time on the benchmark for the larger sizes but it's like 7x worse for the smaller sizes.  I don't think it's any wonder the "byfour" isn't using it to incorporate the running crc into its checksum.  

Perf reports:

```
  57.25%  benchmark_zlib  benchmark_zlib        [.] crc32_pclmulqdq
  17.62%  benchmark_zlib  benchmark_zlib        [.] crc32_combine
  12.38%  benchmark_zlib  benchmark_zlib        [.] crc32_byfour
   7.84%  benchmark_zlib  libc.so.6             [.] random
   2.51%  benchmark_zlib  benchmark_zlib        [.] crc32_pclmulqdq_Benchmark::BenchmarkCase
```

Reply:
That's actually pretty straightforward. I should be able to do this without too much rework.

The stores could be isolated to a single block/label that gets skipped over maybe, it just depends on whether or not register pressure imposes that things end up spilling onto stack or not as to whether we lose the efficiency of the ancillary copy.

As far as splitting up the disparate shared stages, some of that is ideally inlined as the function level abstraction otherwise calls for pushing quite a few things on and off the stack where we'd be better off shuffling through registers.  Apart from maybe the reset function I believe that is what's happening. So as far as efficiency I don't believe it's costing us much to have them broken out like that, but I'll need to take a closer look at a profile.

Reply:
So XOR is commutative, no?  Why are they constantly trying to XOR the very first term read from the buffer (with a goofy branchy do-once macro) when they could just as easily do it one time in the initial first term of the CRC?

Reply:
Hmm, it would appear that the carryless multiplication operations that occur inbetween the XORs won't commute.  I really do need to do this on the very first term and put it everywhere there's a branch for a possible first load.  I mean, I guess that's doable...

Reply:
Well chromium version only does a single xor at the beginning of the function.. 
https://chromium.googlesource.com/chromium/src/third_party/zlib/+/refs/heads/main/crc32_simd.c#48

Reply:
I think it only does that at the beginning, but due to different values of `len`, we don't know where the first read will happen at..?

Reply:
So, not _exactly_ a direct comparison as the first version didn't incorporate the initial CRC value, but here's a comparison with a version that doesn't have to branch at all and bypasses the stores by duplicating the whole function (sans store):
```
Comparing crc32/pclmulqdq (from ./benchmark_zlib_before) to crc32/pclmulqdq (from ./benchmark_zlib)
Benchmark                                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------------------------
[crc32/pclmulqdq vs. crc32/pclmulqdq]/64_mean                       +0.4669         +0.4669            12            18            12            18
[crc32/pclmulqdq vs. crc32/pclmulqdq]/64_median                     +0.4671         +0.4671            12            18            12            18
[crc32/pclmulqdq vs. crc32/pclmulqdq]/64_stddev                     -0.5437         -0.5969             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/64_cv                         -0.6890         -0.7252             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/512_mean                      +0.0307         +0.0307            34            35            34            35
[crc32/pclmulqdq vs. crc32/pclmulqdq]/512_median                    +0.0307         +0.0307            34            35            34            35
[crc32/pclmulqdq vs. crc32/pclmulqdq]/512_stddev                    -0.9257         -0.9234             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/512_cv                        -0.9279         -0.9256             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4096_mean                     -0.0041         -0.0041           175           174           175           174
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4096_median                   -0.0036         -0.0036           175           174           175           174
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4096_stddev                   -0.9845         -0.9860             1             0             1             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4096_cv                       -0.9844         -0.9859             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/32768_mean                    -0.0045         -0.0045          1277          1271          1277          1271
[crc32/pclmulqdq vs. crc32/pclmulqdq]/32768_median                  -0.0047         -0.0048          1277          1271          1277          1271
[crc32/pclmulqdq vs. crc32/pclmulqdq]/32768_stddev                  -0.9223         -0.9268             1             0             1             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/32768_cv                      -0.9219         -0.9264             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/262144_mean                   -0.0004         -0.0004         10034         10030         10034         10030
[crc32/pclmulqdq vs. crc32/pclmulqdq]/262144_median                 -0.0004         -0.0004         10034         10030         10034         10030
[crc32/pclmulqdq vs. crc32/pclmulqdq]/262144_stddev                 +1.3674         +0.9789             0             1             0             1
[crc32/pclmulqdq vs. crc32/pclmulqdq]/262144_cv                     +1.3684         +0.9797             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/2097152_mean                  -0.0989         -0.0989        119229        107441        119226        107438
[crc32/pclmulqdq vs. crc32/pclmulqdq]/2097152_median                -0.0968         -0.0968        119278        107734        119275        107731
[crc32/pclmulqdq vs. crc32/pclmulqdq]/2097152_stddev                +1.2576         +1.2594           511          1153           511          1154
[crc32/pclmulqdq vs. crc32/pclmulqdq]/2097152_cv                    +1.5053         +1.5073             0             0             0             0
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4194304_mean                  -0.0984         -0.0984        238786        215286        238780        215281
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4194304_median                -0.0982         -0.0982        238520        215106        238515        215098
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4194304_stddev                -0.0913         -0.0921          2375          2158          2377          2158
[crc32/pclmulqdq vs. crc32/pclmulqdq]/4194304_cv                    +0.0079         +0.0070             0             0             0             0
```

Reply:
> Well chromium version only does a single xor at the beginning of the function.. https://chromium.googlesource.com/chromium/src/third_party/zlib/+/refs/heads/main/crc32_simd.c#48

Right, I no doubt only need to do it once but apparently I can't do it to values inside the "fold" buffer as that won't properly commute.  The carryless multiplication compounded with the shuffle/lookup table that gets used must be an operation that can't actually commute the value as though it were a simple multiplication or addition on the first set of bytes.

Reply:
Can you add:
```
Copyright (c) 2016 Marian Beermann (support for initial value)
```
From the implementation I mentioned, to the top?

Reply:
done